In [79]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [80]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv("Resources/crypto_market_data.csv", index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [81]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [82]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [83]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
df_market_data_scaled = StandardScaler().fit_transform(df_market_data[["price_change_percentage_24h",
                                                                       "price_change_percentage_7d",
                                                                       "price_change_percentage_14d",
                                                                       "price_change_percentage_30d",
                                                                       "price_change_percentage_60d",
                                                                       "price_change_percentage_200d",
                                                                       "price_change_percentage_1y"]])

# Display the first 3 rows of the scaled data
df_market_data_scaled[0:3]


array([[ 0.50852937,  0.49319307,  0.77220043,  0.23545963, -0.0674951 ,
        -0.35595348, -0.25163688],
       [ 0.18544589,  0.93444504,  0.55869212, -0.05434093, -0.27348273,
        -0.11575947, -0.19935211],
       [ 0.02177396, -0.70633685, -0.02168042, -0.06103015,  0.00800452,
        -0.55024692, -0.28206051]])

In [84]:
# Create a DataFrame with the scaled data
df_market_data_scaled = pd.DataFrame(df_market_data_scaled, columns=["price_change_percentage_24h",
                                                                     "price_change_percentage_7d",
                                                                     "price_change_percentage_14d",
                                                                     "price_change_percentage_30d",
                                                                     "price_change_percentage_60d",
                                                                     "price_change_percentage_200d",
                                                                     "price_change_percentage_1y"])
df_market_data_scaled

# Copy the crypto names from the original data
df_market_data_scaled["coin_id"] = df_market_data.index

# Set the coinid column as index
df_market_data_scaled = df_market_data_scaled.set_index("coin_id")


# Display sample data
df_market_data_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [85]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 11))

In [86]:
# Create an empty list to store the inertia values
inertia = []

In [87]:
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k_values:
    model = KMeans(n_clusters=i, random_state=0)
    model.fit(df_market_data_scaled)
    inertia.append(model.inertia_)


c:\Users\rober\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [88]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_coin = {
    "k": k_values,
    "inertia": inertia
}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_coin = pd.DataFrame(elbow_data_coin)


In [140]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_coin_data_plot = df_elbow_coin.hvplot.line(
    x="k",
    y="inertia",
    title="Coin Data Elbow Curve",
    xticks=k_values)
elbow_coin_data_plot


:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** Based on this Elbow Curve, `k=4` looks like the correct one.

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [90]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=0)

# Print the model
model

KMeans(n_clusters=4, random_state=0)

In [91]:
# Fit the K-Means model using the scaled data
model.fit(df_market_data_scaled)

KMeans(n_clusters=4, random_state=0)

In [92]:
# Predict the clusters to group the cryptocurrencies using the scaled data
coin_movement = model.predict(df_market_data_scaled)

# Print the resulting array of cluster values.
print(coin_movement)

[0 0 3 3 0 0 0 0 0 3 3 3 3 0 3 0 3 3 0 3 3 0 3 3 3 3 3 3 0 3 3 3 1 0 3 3 2
 3 3 3 3]


In [93]:
# Create a copy of the DataFrame
df_coin_predictions = df_market_data.copy()


In [94]:
# Add a new column to the DataFrame with the predicted clusters
df_coin_predictions["Coin Movement Cluster"] = coin_movement

# Display sample data
df_coin_predictions.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                           1.08388                     7.60278   
ethereum                          0.22392                    10.38134   
tether                           -0.21173                     0.04935   
ripple                           -0.37819                    -0.60926   
bitcoin-cash                      2.90585                    17.09717   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                           6.57509                      7.67258   
ethereum                          4.80849                      0.13169   
tether                            0.00640                     -0.04237   
ripple                            2.24984                      0.23455   
bitcoin-cash                     14.75334                     15.74903   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                          -3.25185                      83.51840   
ethereum                        -12.88890                     186.77418   
tether                            0.28037                      -0.00542   
ripple                          -17.55245                      39.53888   
bitcoin-cash                    -13.71793                      21.66042   

              price_change_percentage_1y  Coin Movement Cluster  
coin_id                                                          
bitcoin                         37.51761                      0  
ethereum                       101.96023                      0  
tether                           0.01954                      3  
ripple                         -16.60193                      3  
bitcoin-cash                    14.49384                      0

In [143]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
coin_scatter = df_coin_predictions.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="Coin Movement Cluster",
    hover_cols="coin_id"
)
coin_scatter

:NdOverlay   [Coin Movement Cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [96]:
# Create a PCA model instance and set `n_components=3`.
pca=PCA(n_components=3)

In [97]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
coin_pca = pca.fit_transform(df_market_data_scaled)

# View the first five rows of the DataFrame.
coin_pca[:5] 


array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [98]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_


array([0.3719856 , 0.34700813, 0.17603793])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** `The total explained variance is 37.19% + 34.70% + 17.60%, or almost 90%.  So, the three components combined describe almost 90% of the total variation of the original data.`

In [99]:
# Create a new DataFrame with the PCA data.
coin_pca_df = pd.DataFrame(data=coin_pca, 
                           columns=['PC1', 'PC2', 'PC3'])

# Copy the crypto names from the original data
coin_pca_df["coin_id"] = df_market_data.index

# Set the coinid column as index
coin_pca_df = coin_pca_df.set_index("coin_id")

# Display sample data
coin_pca_df.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [111]:
# Create a list with the number of k-values from 1 to 11
k_values_pca = list(range(1, 11))


In [113]:
# Create an empy list to store the inertia values
pca_inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for p in k_values_pca:
    k_model = KMeans(n_clusters=p, random_state=0)
    k_model.fit(coin_pca_df)
    pca_inertia.append(k_model.inertia_)


c:\Users\rober\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [116]:
# Create a dictionary with the data to plot the Elbow curve
pca_elbow_data = {"k": k_values_pca, "pca_inertia": pca_inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(pca_elbow_data)
df_elbow_pca

k  pca_inertia
0   1   256.874086
1   2   165.901994
2   3    93.774626
3   4    49.665497
4   5    37.878747
5   6    27.618972
6   7    21.182776
7   8    17.389823
8   9    13.593650
9  10    10.559358

In [133]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_pca_plot = df_elbow_pca.hvplot.line(
    x="k", 
    y="pca_inertia", 
    title="PCA Elbow Curve", 
    xticks=k_values_pca
)
elbow_pca_plot

:Curve   [k]   (pca_inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** ``The best value for k, again, appears to be `k=4`.``


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** `No.  It is the same.`

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [118]:
# Initialize the K-Means model using the best value for k
pca_model = KMeans(n_clusters=4, random_state=0)

In [119]:
# Fit the K-Means model using the PCA data
pca_model.fit(coin_pca_df)

KMeans(n_clusters=4, random_state=0)

In [123]:
# Predict the clusters to group the cryptocurrencies using the PCA data
pca_coin_movement = pca_model.predict(coin_pca_df)

# Print the resulting array of cluster values.
pca_coin_movement


array([0, 0, 3, 3, 0, 0, 0, 0, 0, 3, 3, 3, 3, 0, 3, 0, 3, 3, 0, 3, 3, 0,
       3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 1, 0, 3, 3, 2, 3, 3, 3, 3])

In [129]:
# Create a copy of the DataFrame with the PCA data
df_pca_coin_predictions = coin_pca_df.copy()

# Add a new column to the DataFrame with the predicted clusters
df_pca_coin_predictions["PCA Coin Movement Cluster"] = pca_coin_movement

# Display sample data
df_pca_coin_predictions.head()


PC1       PC2       PC3  PCA Coin Movement Cluster
coin_id                                                              
bitcoin      -0.600667  0.842760  0.461595                          0
ethereum     -0.458261  0.458466  0.952877                          0
tether       -0.433070 -0.168126 -0.641752                          3
ripple       -0.471835 -0.222660 -0.479053                          3
bitcoin-cash -1.157800  2.041209  1.859715                          0

In [142]:
# From Canvas Requirements:
# Create a scatter plot using hvPlot as follows:
# - Set the x-axis as "x=PC1"
# - Set the y-axis as "y=PC2"
 
pca_scatter = df_pca_coin_predictions.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="PCA Coin Movement Cluster",
    hover_cols="coin_id"
)
pca_scatter

:NdOverlay   [PCA Coin Movement Cluster]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [141]:
# Composite plot to contrast the Elbow curves
elbow_scatter_compare = elbow_coin_data_plot + elbow_pca_plot
elbow_scatter_compare

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (pca_inertia)

In [144]:
# Composite plot to contrast the clusters
cluster_compare = coin_scatter + pca_scatter
cluster_compare

:Layout
   .NdOverlay.I  :NdOverlay   [Coin Movement Cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [PCA Coin Movement Cluster]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** 